pyop2/sparsity.pyx

```diff
@@ -146,31 +150,48 @@ cdef inline void add_entries_extruded(rset, rmap, cset, cmap,
     roffset = rmap.offset
     coffset = cmap.offset
 
-    layers = rmap.iterset.layers
+    layers = rmap.iterset.layers_array
+
+    constant_layers = rmap.iterset.constant_layers
 
     for region in rmap.iteration_region:
+        region_selector = -1
+        reps = 1
         # The rowmap will have an iteration region attached to
         # it specifying which bits of the "implicit" (walking
         # up the column) map we want.  This mostly affects the
         # range of the loop over layers, except in the
         # ON_INTERIOR_FACETS where we also have to "double" up
         # the map.
-        start = 0
-        end = layers - 1
-        reps = 1
         if region.where == "ON_BOTTOM":
-            end = 1
+            region_selector = 1
         elif region.where == "ON_TOP":
-            start = layers - 2
+            region_selector = 2
         elif region.where == "ON_INTERIOR_FACETS":
-            end = layers - 2
+            region_selector = 3
             reps = 2
         elif region.where != "ALL":
             raise RuntimeError("Unhandled iteration region %s", region)
-
         for e in range(nent):
+            if constant_layers:
+                start = layers[0, 0]
+                end = layers[0, 1] - 1
+            else:
+                start = layers[e, 0]
+                end = layers[e, 1] - 1
+            # Always need to subtract this, because the start isn't always zero.
+            # Even though the map indexes into the base of the column.
+            bottom_layer = start
+            # Adjust start and end appropriately
+            if region_selector == 1:
+                end = start + 1
+            elif region_selector == 2:
+                start = end - 1
+            elif region_selector == 3:
+                end = end - 1
```

Comment:
Agreed, they are just translations to C variables so that the inner loop does not contain python code, can add `#defines`